### ISSUE TITLE
I think "autostart.sh" should not be restricted with the root user check

### ISSUE THREAD

**Issue by oderwat** [janpfeifer/gonb#147]
I think the autostart.sh feature is cool, and we use it in multiple containers now. But I think it should work slightly different. Having GoNB checking it for being owned by root may pose a problem on Windows and is not nice on other systems either. You need to make it UID 0 on the host for that. I am not sure if that even is possible on Windows and I do not like to have a script like that owned by root on the host anyway. 

For added security one could mount that file as read only (`-v ./autostart.sh:/notebooks/autostart.sh:ro`). 

I actually also did not imagine that it is inside "notebooks". I thought it would just be in root and does not show up inside of JupyterLab.

**Comment by thevtm** [comment]
I agree.
In my set up I ended up overriding it using docker-compose.

```yaml
  gonb:
    image: "janpfeifer/gonb_jupyterlab:v0.10.6"
    entrypoint: ""
    command: ["tini", "-g", "--", "/notebooks/docker-compose/gonb/autostart.sh"]
    ports:
      - "58888:8888"
    volumes:
      - ./:/notebooks
    networks:
      - std-network
```

I don't know much about docker but I think we should use `CMD` instead of `ENTRYPOINT` [here](https://github.com/janpfeifer/gonb/blob/0745a3825ad8517a7a696403c013d1753d96ae51/Dockerfile#L105).

**Comment by janpfeifer** [comment]
Thanks for the suggestion, it makes lots of sense.

Just so we are in the same page (or for anyone reading this later), and @oderwat  already referred to it,  the reason we added the check was that it is a relatively easy path for privilege escalation. An "evil" script that manages to run something with user jovyan could simply quietly change `autostart.sh` (which is stored as host, so it outlasts one execution  of the docker), and at the next execution, its injected code would be executed as root (`autostart.sh` is executed as root).

I like @oderwat both proposed ideas:

1. Making the `autostart.sh` readonly. It requires some checking that it was actually mounted readonly.
2. Mounting it under `/root/autostart` -- I should have thought of that, there is no need to be in the `/notebooks` directory indeed. Since it is a directory, the file should be under `/root/autostart/autostart.sh`.

Would that be ok ? I'll create a new docker version today or tomorrow.

Hi @thevtm , I also imagined someone overwriting the `ENTRYPOINT` (by using the `--entrypoint` flag when executing the docker) as means to run some arbitrary setup script. But still, the `autostart.sh` is convenient in some cases.

I'm no docker expert either, but I remember reading the documentation on `ENTRYPOINT` and `CMD`, and I think `ENTRYPOINT` (the default executable) is right thing to set up. `CMD` being the default argument that can be easily changed. Ask GPT / Gemini or any other LLM, it will have a good explanation.

**Comment by oderwat** [comment]
I this that 1/2 makes sense. I believe that you need to attempt to write to the file to check if it is RO, because the flags will not show that. Likewise, I wonder if that really should be done. Having the "autostart" directory 0700 should be enough and if somebody wants he could mount over it?

A bit off-topic but I think it can be interesting anyway:

About ENTRYPOINT and CMD I was just two days ago doing exactly that: Discussing the topic with ClaudeAI. As I finally understood that ENTRYPOINT is getting CMD as "arguments" and these are also the ones you put after the image name, it clicked for me. One other difference is that you can not use ARG in the ENTRYPOINT.

As I understood that I can overwrite the ENTRYPOINT of any docker file I remembered older problems of mine: Listing the files in an docker image, extracting files or directories from it and getting diffs between image versions. So I was continuing the ClaudeAI conversation and "we" made this in the end: [docker-inspector](https://github.com/oderwat/docker-inspector).

**Comment by janpfeifer** [comment]
Ok, I ended up implementing with the `readonly` mount requirement. Indeed, if someone mounts something under root and not writable to others should be safe enough ... but just in case -- it's easy to check, I just need to check I can't chmod the file to its current permissions (so not changing anything). 

It seems to be working, and I pushed a new docker tagged `janpfeifer/gonb_jupyterlab:latest` or `janpfeifer/gonb_jupyterlab:v0.10.6-20241102` to [DockerHub](https://hub.docker.com/r/janpfeifer/gonb_jupyterlab). 

Pls, let me know if you find any other issues, and thanks again for the suggestion!

ps1.: Nicely done with `docker-inspector`. I'll keep that in mind for when I need to manipulate/compare dockers.

ps2.: Apologies for the delay, this week I've been intensely coding on [another project on converting ONNX models](https://github.com/gomlx/onnx-gomlx) to [GoMLX](https://github.com/gomlx/gomlx), which just started to work with larger models and I"m super happy about 😃 -- in case you are interested in ML projects in Go.
